capstone-backend/src/main/java/com/google/sps/servlets/BookServlet.java

```diff
@@ -0,0 +1,138 @@
+package com.google.sps.servlets;
+
+import com.google.api.core.ApiFuture;
+import com.google.cloud.firestore.CollectionReference;
+import com.google.cloud.firestore.DocumentReference;
+import com.google.cloud.firestore.DocumentSnapshot;
+import com.google.cloud.firestore.Firestore;
+import com.google.cloud.firestore.QueryDocumentSnapshot;
+import com.google.cloud.firestore.QuerySnapshot;
+import com.google.cloud.firestore.SetOptions;
+import com.google.cloud.firestore.WriteResult;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+import com.google.sps.data.Book;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.HashMap;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+
+@WebServlet("/api/books")
+public class BookServlet extends HttpServlet {
+  private Firestore db; 
+  private CollectionReference books;
+  private Gson gson;
+
+  public BookServlet() throws IOException {
+    db = Utility.getFirestoreDb();
+    books = db.collection("books");
+    gson = new Gson();
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    if (request.getParameter("id") != null) {
+      DocumentReference docRef = books.document(request.getParameter("id"));
+      ApiFuture<DocumentSnapshot> future = docRef.get();
```

Comment:
Have you tried testing (locally is fine) what happens when you make a GET request for some "id" that doesn't exist in the database (or if the database is empty)? Do you get the `"Error: no such document!"` case?

Reply:
Yes you do in that case. In the documentation for Firestore it says "if there is no document at the location referenced by docRef, the resulting document will be empty and calling exists on it will return false," (check it out here: https://firebase.google.com/docs/firestore/query-data/get-data#get_a_document) which is what I based this on. 

I figured the best way to handle this would just be printing to System.err and returning a null book, but can change that up if there's a better way.